doc/toolchains.md

```diff
@@ -0,0 +1,80 @@
+or may have not selected a default toolchain for use via `alr toolchain
+--select`. The interaction between these two features is explained next.
+
+Before going into the details, there are two kind of dependencies on GNAT
+compilers: generic dependencies on the `gnat` crate, which apply to every 
+compiler, and dependencies on a precise native or cross-compiler, e.g.,
+`gnat_native` or `gnat_riscv_elf`.
+
+## Identifying available compilers
+
+Available compilers can be listed with `alr search --full --external-detect
+gnat_`. They will also be shown by the selection assistant, `alr toolchain
+--select`.
+
+## Specifying dependencies on GNAT compilers
+
+From the point of view of a user preparing a release for publication, these
+are the considerations to bear in mind:
+
+- Do not use any dependency on a compiler if you do not have a reason to do so.
+- Use dependencies on `gnat` to specify known compiler version restrictions.
+- Use dependencies on a precise gnat cross-compiler (e.g., `gnat_riscv_elf`)
+  when your crate is purposely only available for such a target.
+- There is no reason to specify a dependency on the native compiler
+  (`gnat_native`) as that would unnecessarily limit the availability of a
+  library crate that might be useful to dependent cross-target crates.
+
+## Interactions with a selected toolchain
+
+From the point of view of a user wanting to compile some release, there will be
+an interaction between the solution dependency on compilers and a selected
+default compiler. 
+
+The simplest and most usual scenario for native compilation is as follows:
+
+- Solution without dependencies on `gnat`: 
+   - If a default compiler has been configured: the compiler will be included in the
+     PATH provided by `alr printenv` and used for the build.
+   - No compiler has been configured: `alr` will attempt to use whatever toolchain is
+     available in the user's environment prior to running `alr`. 
+      - Note that compilers deployed with `alr toolchain --install` but not
+        selected as the default will not be used in this case.
+
+For crates with dependencies on GNAT compilers, the following two cases simultaneously apply:
+
+- Solution with generic dependencies on `gnat`: The solver will provide a
+  concrete GNAT to satisfy this dependency, applying the following
+  prioritization:
+   1. A target-specific compiler that is already a dependency.
+   1. The default compiler, if it has been defined.
+   1. A native compiler that is already deployed.
+   1. A cross-compiler that is already deployed.
+   1. A native compiler from the index, that will be deployed with the rest of dependencies.
+   1. A cross-compiler from the index, that will be deployed with the rest of dependencies.
+
+- Solution with dependencies on a target-specific GNAT (`gnat_native`, `gnat_ricv_elf`,
+  etc.): The `alr` solver will provide one compiler in the environment, applying the
```

Comment:
Simultaneously in a crate? How's that supposed to work?

Reply:
I have one video game crate (`shoot_n_loot`) that builds both the embedded version of the game and a native simulator to play it on the computer. But maybe that's a very bad idea ^^

Reply:
I actually was worried about this use case but I guess I didn't traslate it properly in our talks. This is the one where I saw your idea on conditional dependencies on something provided via command line/dependent crate configuration could be useful:
```
alr build target:=native  # arbitrary name, not predefined (or defined in the configuration section of the crate)
```
and then we can have in the toml:
```
[[depends-on."case(crate.target)".native]]
gnat_native = "*"
[[depends-on."case(crate.target)".riscv]]
gnat_riscv_elf = "*"
```

There should be a default somehow for automated builds though.

For a later date obviously...

Otherwise, I guess we can do an exception about the `gnat` crate not being conflicting among compilers. But something nags me at that being unkind to people not interested in all targets.

Reply:
Let's keep it simple for now and only allow one compiler.

Reply:
I have one other use case. 

A cross project that uses native tools to generate code, for instance ![https://alire.ada.dev/crates/are](https://img.shields.io/endpoint?url=https://alire.ada.dev/badges/are.json).

Reply:
Right, that is a legitimate situation that we ought to support. I'll try to come up with a solution in a separate PR.

Reply:
Addressed in https://github.com/alire-project/alire/pull/780